README.md
# cookie-store

---

[![Build Status](https://travis-ci.org/MoeKit/cookieStorage.svg)](https://travis-ci.org/MoeKit/cookieStorage)
[![Coverage Status](https://coveralls.io/repos/MoeKit/cookieStorage/badge.png)](https://coveralls.io/r/MoeKit/cookieStorage?branch=moe)

---

避免浪费多个cookie名，多个前端`cookie`值保存到一个cookie里

当命名空间为`fe_`时，所有cookie保存在`fe_0`为名的cookie里，为`fe_0`保存的值达到浏览器允许字符数时，会自动创建`fe_1`的cookie.

## 使用说明

站内使用，请手动设置cookie保存域名，全域为 `seedit.com`

````javascript
define(function(require){
    var cookieStore = require('index');
    var test = new cookieStore('fe_',{
        domain: 'seedit.com',
        expires: 365
    });
    // 删除
    test.remove('aa');
    // 获取
    console.log(test.get('aa'));
    // 保存
    test.put('aa','bb');
    console.log(test.get('aa'));
});
````

## 注意

+ 能不用cookie就不用cookie
+ 能不设置到全域就不设置到全域
+ 可以设置到特定path就不要设置到'/'
+ 前端cookie命名空间为 `fe_`
+ 尽量在需求下线时清除掉已经设定的cookie


HISTORY.md
# History

---

## 0.0.9

`CHANGED` spm@3.x 规范

## 0.0.8

`tag:new` moe/cookieStore 初次提交



tests/cookieStore-spec.js
var ObjProto = Object.prototype ,
    hasOwnProperty = ObjProto.hasOwnProperty,
    isObject = function (obj) {
        return obj === Object(obj);
    },
    has = function (obj, key) {
        return hasOwnProperty.call(obj, key);
    },
    nativeKeys = Object.keys,
    getKeys = function (obj) {
        if (!isObject(obj)) return [];
        if (nativeKeys) return nativeKeys(obj);
        var keys = [];
        for (var key in obj) if (has(obj, key)) keys.push(key);
        return keys;
    };


var each = function (obj, iterator, context) {
    var ArrayProto = Array.prototype;
    var nativeForEach = ArrayProto.forEach;
    var breaker = {};
    if (obj == null) return obj;
    if (nativeForEach && obj.forEach === nativeForEach) {
        obj.forEach(iterator, context);
    } else if (obj.length === +obj.length) {
        for (var i = 0, length = obj.length; i < length; i++) {
            if (iterator.call(context, obj[i], i, obj) === breaker) return;
        }
    } else {
        var keys = getKeys(obj);
        for (var i = 0, length = keys.length; i < length; i++) {
            if (iterator.call(context, obj[keys[i]], keys[i], obj) === breaker) return;
        }
    }
    return obj;
};

var CookieStorage = (function () {
    var CookieManager;

    CookieManager = (function () {
        var set, get, remove, isSet, encode;

        encode = function (s) {
            s = s.replace(/,/g, '%2C');
            s = s.replace(/;/g, '%3B');
            s = s.replace(/\s/g, '%20');
            s = s.replace(/\|/g, '%7C');

            return s;
        };

        isSet = function (name) {
            return (new RegExp("(?:^|;\\s*)" + encodeURI(name).replace(/[\-\.\+\*]/g, "\\$&") + "\\s*\\=")).test(document.cookie);
        };

        set = function (name, value, option) {
            option = option || {};
            var cookie = encodeURI(name) + "=" + value + ";path=/";
            if (option.domain) {
                cookie += ';domain=' + option.domain;
            }
            if (option.expires) {
                var expires = new Date((+new Date()) + option.expires * 24 * 60 * 60 * 1000).toGMTString();
                cookie += ';expires=' + expires;
            }
            document.cookie = cookie;
        };

        get = function (name) {
            if (!isSet(name)) {
                return null;
            }
            return decodeURIComponent(document.cookie.replace(new RegExp("(?:^|.*;\\s*)" + encodeURI(name).replace(/[\-\.\+\*]/g, "\\$&") + "\\s*\\=\\s*((?:[^;](?!;))*[^;]?).*"), "$1"));
        };

        remove = function (name) {
            var oExpDate = new Date();
            oExpDate.setDate(oExpDate.getDate() - 1);
            document.cookie = encodeURI(name) + "=expired;expires=" + oExpDate.toGMTString() + ";path=/";
        };

        return {
            set: set,
            get: get,
            remove: remove,
            isSet: isSet,
            encode: encode
        };

    }());


    return CookieManager

}());


var CookieStore = require('index');
var expect = require('expect');
var _ = require('underscore');
describe('cookie-storage', function () {
    var cs = new CookieStore('cskv_'),
        cm = CookieStorage,
        clearCookies,
        getRandomArbitary,
        randomWords;

    clearCookies = function () {
        each(document.cookie.split(';'), function (cookie) {
            var name = cookie.split('='),
                oExpDate = new Date();

            oExpDate.setDate(oExpDate.getDate() - 1);
            document.cookie = escape(name[0]) + "=; expires=" + oExpDate.toGMTString() + "; path=/";
        });
    };

    getRandomArbitary = function (min, max) {
        return Math.floor(Math.random() * (max - min) + min);
    };

    randomWords = function (count) {
        var letters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ',
            minSize = 3,
            maxSize = 16,
            result = '';

        _(count).times(function (n) {
            _(getRandomArbitary(minSize, maxSize)).times(function () {
                result += letters[getRandomArbitary(0, letters.length)];
            });

            if (n < count - 1) {
                result += ' ';
            }
        });

        return result;
    };

    beforeEach(function () {
        clearCookies();
    });

    afterEach(function () {
        clearCookies();
    });

    /*
     * CookieStorage.CookieManger
     */
    describe('cookie-manager', function () {
        it('should return null', function () {
            expect(cm.get('does_not_exist')).to.be(null);
        });

        it('should get what was just set', function () {
            var varName = 'shouldEquals',
                varValue = 'a test string ? ; = ,';
            cm.set(varName, cm.encode(varValue));

            expect(cm.get(varName)).to.be(varValue);
        });

        it('should remove the cookie', function () {
            var name = "cookieName",
                value = "value";

            cm.set(name, value);
            expect(cm.get(name)).to.be(value);

            cm.remove(name);
            expect(cm.get(name)).to.be(null);
        });

        it('should pass when set, fail when not set', function () {
            var name = "cookieName",
                value = "value";

            cm.set(name, value);
            expect(cm.isSet(name)).to.be(true);

            cm.remove(name);
            expect(cm.isSet(name)).to.be(false);
        });
    });

    /*
     * CookieStorage
     */
    it('should have a cookie size of 0', function () {
        expect(document.cookie.length).to.be(0);
    });

    it('should "get" same data that was "put"', function () {
        var name = "cookieName",
            value = "value";

        cs.put(name, value);
        expect(cs.get(name)).to.be(value);
    });

    it('should have an expected length', function () {
        var name = "name",
            value = "her royal majesty queen elizabeth ii";
        cs.put(name, value);
        expect(cs.entrySize('name')).to.be(51);
    });

    it('should have a no cookie stores', function () {
        cs.clearStores();
        cs.loadStores();
        expect(cs.listStores().length).to.be(0);
    });

    it('should have a single cookie store', function () {
        var name = "cookieName",
            value = "value";
        cs.loadStores();
        expect(cs.listStores().length).to.be(0);
        cs.put(name, value);
        expect(cs.listStores().length).to.be(1);
    });

    it('should remove the specified cookie', function () {
        cs.put('test', 'testvalue');
        expect(cs.get('test')).to.be('testvalue');
        cs.remove('test');
        expect(cs.get('test')).to.be(undefined);
    });

    it('should span multiple cookie stores', function () {
        var numEntries = 80;

        cs.loadStores();

        _(numEntries).times(function () {
            cs.put(randomWords(1), randomWords(getRandomArbitary(1, 11)));
        });

        expect(cs.listStores().length > 1).to.be(true);
    });
});




spec/cookie-storage-spec.js
console.log(_);
describe('cookie-storage', function () {
    var cs = new CookieStore('cskv_'),
        cm = CookieStorage.CookieManager,
        clearCookies,
        getRandomArbitary,
        randomWords;

    clearCookies = function () {
        _.each(document.cookie.split(';'), function (cookie) {
            var name = cookie.split('='),
                oExpDate = new Date();

            oExpDate.setDate(oExpDate.getDate() - 1);
            document.cookie = escape(name[0]) + "=; expires=" + oExpDate.toGMTString() + "; path=/";
        });
    };

    getRandomArbitary = function (min, max) {
        return Math.floor(Math.random() * (max - min) + min);
    }

    randomWords = function (count) {
        var letters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ',
            minSize = 3,
            maxSize = 16,
            result = '';

        _(count).times(function (n) {
            _(getRandomArbitary(minSize, maxSize)).times(function () {
                result += letters[getRandomArbitary(0, letters.length)];
            });

            if (n < count - 1) {
                result += ' ';
            }
        });

        return result;
    };

    beforeEach(function () {
        clearCookies();
    });

    afterEach(function(){
        clearCookies();
    });

    /*
     * CookieStorage.CookieManger
     */
    describe('cookie-manager', function () {
        it('should return null', function () {
            expect(cm.get('does_not_exist')).toEqual(null);
        });

        it('should get what was just set', function () {
            var varName = 'shouldEquals',
                varValue = 'a test string ? ; = ,';
            cm.set(varName, cm.encode(varValue));

            expect(cm.get(varName)).toEqual(varValue);
        });

        it('should remove the cookie', function () {
            var name = "cookieName",
                value = "value";

            cm.set(name, value);
            expect(cm.get(name)).toEqual(value);

            cm.remove(name);
            expect(cm.get(name)).toEqual(null);
        });

        it('should pass when set, fail when not set', function () {
            var name = "cookieName",
                value = "value";

            cm.set(name, value);
            expect(cm.isSet(name)).toEqual(true);

            cm.remove(name);
            expect(cm.isSet(name)).toEqual(false);
        });
    });

    /*
     * CookieStorage
     */
    it('should have a cookie size of 0', function () {
        expect(document.cookie.length).toEqual(0);
    });

    it('should "get" same data that was "put"', function () {
        var name = "cookieName",
            value = "value";

        cs.put(name, value);
        expect(cs.get(name)).toEqual(value);
    });

    it('should have an expected length', function () {
        var name = "name",
            value = "her royal majesty queen elizabeth ii";
        cs.put(name, value);
        expect(cs.entrySize('name')).toEqual(51);
    });

    it('should have a no cookie stores', function () {
        cs.clearStores();
        cs.loadStores();
        expect(cs.listStores().length).toEqual(0);
    });

    it('should have a single cookie store', function () {
        var name = "cookieName",
            value = "value";
        cs.loadStores();
        expect(cs.listStores().length).toEqual(0);
        cs.put(name, value);
        expect(cs.listStores().length).toEqual(1);
    });

    it('should remove the specified cookie', function () {
        cs.put('test', 'testvalue');
        expect(cs.get('test')).toEqual('testvalue');
        cs.remove('test');
        expect(cs.get('test')).toEqual(undefined);
    });

    it('should span multiple cookie stores', function () {
        var numEntries = 80;

        cs.loadStores();

        _(numEntries).times(function () {
            cs.put(randomWords(1), randomWords(getRandomArbitary(1, 11)));
        });

        expect(cs.listStores().length > 1).toEqual(true);
    });
});

Gruntfile.js
module.exports = function(grunt) {
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        jasmine: {
            main: {
                src: 'src/cookie-storage.js',
                options: {
                    specs: 'spec/cookie-storage-spec.js',
                    helpers: 'http://underscorejs.org/underscore-min.js'
                }
            }
        },
        uglify: {
            options: {
                banner: '/*! <%= pkg.name %> v<%= pkg.version %> | <%= pkg.license %> */\n'
            },
            build: {
                src: 'src/cookie-storage.js',
                dest: 'dist/cookieStorage.min.js'
            }
        }
    });

    grunt.loadNpmTasks('grunt-contrib-jasmine');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.registerTask('default', ['jasmine:main', 'uglify']);
    grunt.registerTask('ci', ['default']);
}

src/cookieStore.js
var ObjProto = Object.prototype ,
    hasOwnProperty = ObjProto.hasOwnProperty,
    isObject = function (obj) {
        return obj === Object(obj);
    },
    has = function (obj, key) {
        return hasOwnProperty.call(obj, key);
    },
    nativeKeys = Object.keys,
    getKeys = function (obj) {
        if (!isObject(obj)) return [];
        if (nativeKeys) return nativeKeys(obj);
        var keys = [];
        for (var key in obj) if (has(obj, key)) keys.push(key);
        return keys;
    };


var each = function (obj, iterator, context) {
    var ArrayProto = Array.prototype;
    var nativeForEach = ArrayProto.forEach;
    var breaker = {};
    if (obj == null) return obj;
    if (nativeForEach && obj.forEach === nativeForEach) {
        obj.forEach(iterator, context);
    } else if (obj.length === +obj.length) {
        for (var i = 0, length = obj.length; i < length; i++) {
            if (iterator.call(context, obj[i], i, obj) === breaker) return;
        }
    } else {
        var keys = getKeys(obj);
        for (var i = 0, length = keys.length; i < length; i++) {
            if (iterator.call(context, obj[keys[i]], keys[i], obj) === breaker) return;
        }
    }
    return obj;
};


var CookieStorage = CookieStorage || (function () {
    var CookieManager;

    CookieManager = (function () {
        var set, get, remove, isSet, encode;

        encode = function (s) {
            s = s.replace(/,/g, '%2C');
            s = s.replace(/;/g, '%3B');
            s = s.replace(/\s/g, '%20');
            s = s.replace(/\|/g, '%7C');

            return s;
        };

        isSet = function (name) {
            return (new RegExp("(?:^|;\\s*)" + encodeURI(name).replace(/[\-\.\+\*]/g, "\\$&") + "\\s*\\=")).test(document.cookie);
        };

        set = function (name, value, option) {
            option = option || {};
            var cookie = encodeURI(name) + "=" + value + ";path=/";
            if (option.domain) {
                cookie += ';domain=' + option.domain;
            }
            if (option.expires) {
                var expires = new Date((+new Date()) + option.expires * 24 * 60 * 60 * 1000).toGMTString();
                cookie += ';expires=' + expires;
            }
            document.cookie = cookie;
        };

        get = function (name) {
            if (!isSet(name)) {
                return null;
            }
            return decodeURIComponent(document.cookie.replace(new RegExp("(?:^|.*;\\s*)" + encodeURI(name).replace(/[\-\.\+\*]/g, "\\$&") + "\\s*\\=\\s*((?:[^;](?!;))*[^;]?).*"), "$1"));
        };

        remove = function (name) {
            var oExpDate = new Date();
            oExpDate.setDate(oExpDate.getDate() - 1);
            document.cookie = encodeURI(name) + "=expired;expires=" + oExpDate.toGMTString() + ";path=/";
        };

        return {
            set: set,
            get: get,
            remove: remove,
            isSet: isSet,
            encode: encode
        };

    }());

    return {
        CookieManager: CookieManager
    };
}());

var CookieStore = CookieStore || (function () {
    var CookieManager, MAX_COOKIE_LENGTH;

    MAX_COOKIE_LENGTH = 4096;
    CookieManager = CookieStorage.CookieManager;

    function CookieStore(namespace, cookieOption) {
        this.prefix = namespace;
        this.cookieOption = cookieOption;
        this.loadStores();
    }

    CookieStore.prototype.persist = function () {
        var vCookieKey,
            vCookieVal,
            index = 0,
            vCookie = {},
            store = this;

        each(this.kvstore, function (value, key) {
            vCookieKey = store.prefix + index;
            vCookieVal = vCookie[vCookieKey] || '';

            if (store.cookieSize(vCookieVal) + store.entrySize(key) > MAX_COOKIE_LENGTH) {
                index += 1;
            }

            vCookieVal += key + '=' + CookieManager.encode(value) + '|';
            vCookie[vCookieKey] = vCookieVal;
        });

        store.clearStores();

        each(vCookie, function (value, key) {
            CookieManager.set(key, value, store.cookieOption);
        });
    };

    CookieStore.prototype.listStores = function () {
        var regEx = new RegExp(this.prefix + "\\d+", "g");
        return document.cookie.match(regEx) || [];
    };

    CookieStore.prototype.clearStores = function () {
        each(this.listStores(), function (storeKey) {
            CookieManager.remove(storeKey);
        });
    };

    CookieStore.prototype.loadStores = function () {
        var cookieVal, parts,
            kvo = {};

        each(this.listStores(), function (storeKey) {
            cookieVal = CookieManager.get(storeKey);
            each(cookieVal.split('|'), function (pair) {
                parts = pair.split("=");
                if (parts[0]) {
                    kvo[parts[0]] = parts[1];
                }
            });
        });

        this.kvstore = kvo;
    };

    /**
     * k=val
     * k.length + '='.length + kvstore[k].length
     */
    CookieStore.prototype.entrySize = function (k) {
        return k.length + 1 + (this.kvstore[k] ? CookieManager.encode(this.kvstore[k]).length : 0);
    };

    CookieStore.prototype.cookieSize = function (value) {
        return this.prefix.length + 2 + (value ? CookieManager.encode(value).length : 0);
    };

    CookieStore.prototype.put = function (k, v) {
        this.kvstore[k] = v;
        this.persist();
    };

    CookieStore.prototype.get = function (k) {
        return this.kvstore[k];
    };

    CookieStore.prototype.remove = function (k) {
        delete(this.kvstore[k]);
        this.persist();
    };

    return CookieStore;
})();

module.exports = CookieStore;


src/cookie-storage.js
/*jslint nomen: true, indent: 2 */
/*global _:true, document:true */
(function(scope) {
    "use strict";

    var each = function(obj, iterator, context) {
        var ArrayProto = Array.prototype;
        var nativeForEach = ArrayProto.forEach;
        var breaker = {};
        if (obj == null) return obj;
        if (nativeForEach && obj.forEach === nativeForEach) {
            obj.forEach(iterator, context);
        } else if (obj.length === +obj.length) {
            for (var i = 0, length = obj.length; i < length; i++) {
                if (iterator.call(context, obj[i], i, obj) === breaker) return;
            }
        } else {
            var keys = _.keys(obj);
            for (var i = 0, length = keys.length; i < length; i++) {
                if (iterator.call(context, obj[keys[i]], keys[i], obj) === breaker) return;
            }
        }
        return obj;
    };

    if (!scope) {
        scope = {};
    }

    scope.CookieStorage = scope.CookieStorage || (function() {
        var CookieManager;

        CookieManager = (function() {
            var set, get, remove, isSet, encode;

            encode = function(s) {
                s = s.replace(/,/g, '%2C');
                s = s.replace(/;/g, '%3B');
                s = s.replace(/\s/g, '%20');
                s = s.replace(/\|/g, '%7C');

                return s;
            };

            isSet = function(name) {
                return (new RegExp("(?:^|;\\s*)" + encodeURI(name).replace(/[\-\.\+\*]/g, "\\$&") + "\\s*\\=")).test(document.cookie);
            };

            set = function(name, value, option) {
                var cookie = encodeURI(name) + "=" + value + ";path=/";
                if (option && option.domain) {
                    cookie += ';domain=' + option.domain;
                }
                if (option && option.expires) {
                    var expires = new Date((+new Date()) + option.expires * 24 * 60 * 60 * 1000).toGMTString();
                    cookie += ';expires=' + expires;
                }
                document.cookie = cookie;
            };

            get = function(name) {
                if (!isSet(name)) {
                    return null;
                }
                return decodeURIComponent(document.cookie.replace(new RegExp("(?:^|.*;\\s*)" + encodeURI(name).replace(/[\-\.\+\*]/g, "\\$&") + "\\s*\\=\\s*((?:[^;](?!;))*[^;]?).*"), "$1"));
            };

            remove = function(name) {
                var oExpDate = new Date();
                oExpDate.setDate(oExpDate.getDate() - 1);
                document.cookie = encodeURI(name) + "=expired;expires=" + oExpDate.toGMTString() + ";path=/";
            };

            return {
                set: set,
                get: get,
                remove: remove,
                isSet: isSet,
                encode: encode
            };

        }());

        return {
            CookieManager: CookieManager
        };
    }());

    scope.CookieStore = scope.CookieStore || (function() {
        var CookieManager, MAX_COOKIE_LENGTH;

        MAX_COOKIE_LENGTH = 4096;
        CookieManager = CookieStorage.CookieManager;

        function CookieStore(namespace, cookieOption) {
            this.prefix = namespace;
            this.cookieOption = cookieOption;
            this.loadStores();
        }

        CookieStore.prototype.persist = function() {
            var vCookieKey,
                vCookieVal,
                index = 0,
                vCookie = {},
                store = this;

            each(this.kvstore, function(value, key) {
                vCookieKey = store.prefix + index;
                vCookieVal = vCookie[vCookieKey] || '';

                if (store.cookieSize(vCookieVal) + store.entrySize(key) > MAX_COOKIE_LENGTH) {
                    index += 1;
                }

                vCookieVal += key + '=' + CookieManager.encode(value) + '|';
                vCookie[vCookieKey] = vCookieVal;
            });

            store.clearStores();

            each(vCookie, function(value, key) {
                CookieManager.set(key, value, store.cookieOption);
            });
        };

        CookieStore.prototype.listStores = function() {
            var regEx = new RegExp(this.prefix + "\\d+", "g");
            return document.cookie.match(regEx) || [];
        };

        CookieStore.prototype.clearStores = function() {
            each(this.listStores(), function(storeKey) {
                CookieManager.remove(storeKey);
            });
        };

        CookieStore.prototype.loadStores = function() {
            var cookieVal, parts,
                kvo = {};

            each(this.listStores(), function(storeKey) {
                cookieVal = CookieManager.get(storeKey);
                each(cookieVal.split('|'), function(pair) {
                    parts = pair.split("=");
                    if (parts[0]) {
                        kvo[parts[0]] = parts[1];
                    }
                });
            });

            this.kvstore = kvo;
        };

        /**
         * k=val
         * k.length + '='.length + kvstore[k].length
         */
        CookieStore.prototype.entrySize = function(k) {
            return k.length + 1 + (this.kvstore[k] ? CookieManager.encode(this.kvstore[k]).length : 0);
        };

        CookieStore.prototype.cookieSize = function(value) {
            return this.prefix.length + 2 + (value ? CookieManager.encode(value).length : 0);
        };

        CookieStore.prototype.put = function(k, v) {
            this.kvstore[k] = v;
            this.persist();
        };

        CookieStore.prototype.get = function(k) {
            return this.kvstore[k];
        };

        CookieStore.prototype.remove = function(k) {
            delete(this.kvstore[k]);
            this.persist();
        };

        return CookieStore;
    })();
}(this));